vm/devices/storage/scsidisk/src/scsidvd/mod.rs

```diff
@@ -2515,6 +2532,100 @@ mod tests {
         scsi_dvd
     }
 
+    struct ScsiDvdNvmeTest {
+        scsi_dvd: SimpleScsiDvd,
+        _nvme_driver: NvmeDriver<EmulatedDevice<NvmeController>>, // We need to store this to keep it from going out of scope
+    }
+
+    impl ScsiDvdNvmeTest {
+        async fn new(
+            driver: DefaultDriver,
+            sector_size: u32,
+            sector_count: u64,
+            read_only: bool,
+        ) -> Self {
+            const MSIX_COUNT: u16 = 2;
+            const IO_QUEUE_COUNT: u16 = 64;
+            const CPU_COUNT: u32 = 64;
+
+            let driver_source = VmTaskDriverSource::new(SingleDriverBackend::new(driver));
+            let base_len = 64 << 20;
+            let payload_len = 4 << 30;
+            let mem = DeviceSharedMemory::new(base_len, payload_len);
+            let mut msi_set = MsiInterruptSet::new();
+            let nvme = NvmeController::new(
+                &driver_source,
+                mem.guest_memory().clone(),
+                &mut msi_set,
+                &mut ExternallyManagedMmioIntercepts,
+                NvmeControllerCaps {
+                    msix_count: MSIX_COUNT,
+                    max_io_queues: IO_QUEUE_COUNT,
+                    subsystem_id: Guid::new_random(),
+                },
+            );
+
+            let payload_mem = mem
+                .guest_memory()
+                .subrange(base_len as u64, payload_len as u64, false)
+                .unwrap();
+            let driver_dma_mem = payload_mem.clone();
+            let mut buf = vec![0u8; sector_count as usize * sector_size as usize];
+            let mut temp = vec![0u8; sector_size as usize];
+            assert!(sector_size > 2);
+            temp[sector_size as usize / 2 - 1] = 2;
+            temp[sector_size as usize / 2] = 3;
+
+            for i in (0..buf.len()).step_by(temp.len()) {
+                let end_point = i + temp.len();
+                buf[i..end_point].copy_from_slice(&temp);
+            }
+            payload_mem.write_at(0, &buf).unwrap();
+
+            nvme.client()
+                .add_namespace(
+                    1,
+                    disklayer_ram::ram_disk(sector_size as u64 * sector_count, read_only).unwrap(),
+                )
+                .await
+                .unwrap();
+
+            let device = EmulatedDevice::new(nvme, msi_set, mem);
+            let nvme_driver = NvmeDriver::new(&driver_source, CPU_COUNT, device)
+                .await
+                .unwrap();
+            let namespace = nvme_driver.namespace(1).await.unwrap();
+            let buf_range = OwnedRequestBuffers::linear(0, 16384, true);
+            namespace
+                .write(
+                    0,
+                    0,
+                    cmp::min(sector_count, 8) as u32, // Limit to a small number of blocks so we can do it all in one operation
```

Comment:
Why is it important that we do it all in one operation?

Reply:
It's not important, but we only look at the first small size anyway, so we don't need to fill the entire disk with a more complex multi-step write here.